File: common/src/main/java/juuxel/adorn/client/gui/widget/WidgetEnvelope.java
package juuxel.adorn.client.gui.widget;

import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.Renderable;
import net.minecraft.client.gui.components.events.GuiEventListener;
import net.minecraft.client.gui.narration.NarratableEntry;
import net.minecraft.client.gui.narration.NarrationElementOutput;
import net.minecraft.client.gui.narration.NarrationSupplier;
import net.minecraft.client.input.CharacterEvent;
import net.minecraft.client.input.KeyEvent;
import net.minecraft.client.input.MouseButtonEvent;

/**
 * A wrapper for a widget (obtained by calling {@link #current()}).
 *
 * <p>Used for enhancing the widget with some functionality,
 * such as {@linkplain FlipBook pagination}, {@linkplain ScissorEnvelope scissoring} or
 * {@linkplain ScrollEnvelope scrolling}.
 */
public abstract class WidgetEnvelope implements GuiEventListener, Renderable, NarratableEntry, TickingElement, Draggable {
    protected abstract GuiEventListener current();

    @Override
    public void extractRenderState(GuiGraphicsExtractor context, int mouseX, int mouseY, float delta) {
        if (current() instanceof Renderable drawable) {
            drawable.extractRenderState(context, mouseX, mouseY, delta);
        }
    }

    @Override
    public void mouseMoved(double mouseX, double mouseY) {
        current().mouseMoved(mouseX, mouseY);
    }

    @Override
    public boolean mouseClicked(MouseButtonEvent click, boolean doubled) {
        return current().mouseClicked(click, doubled);
    }

    @Override
    public boolean mouseReleased(MouseButtonEvent click) {
        return current().mouseReleased(click);
    }

    @Override
    public boolean mouseDragged(MouseButtonEvent click, double offsetX, double offsetY) {
        return current().mouseDragged(click, offsetX, offsetY);
    }

    @Override
    public boolean mouseScrolled(double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
        return current().mouseScrolled(mouseX, mouseY, horizontalAmount, verticalAmount);
    }

    @Override
    public boolean keyPressed(KeyEvent input) {
        return current().keyPressed(input);
    }

    @Override
    public boolean keyReleased(KeyEvent input) {
        return current().keyReleased(input);
    }

    @Override
    public boolean charTyped(CharacterEvent input) {
        return current().charTyped(input);
    }

    @Override
    public boolean isFocused() {
        return current().isFocused();
    }

    @Override
    public void setFocused(boolean focused) {
        current().setFocused(focused);
    }

    @Override
    public boolean isMouseOver(double mouseX, double mouseY) {
        return current().isMouseOver(mouseX, mouseY);
    }

    @Override
    public void updateNarration(NarrationElementOutput builder) {
        if (current() instanceof NarrationSupplier narratable) {
            narratable.updateNarration(builder);
        }
    }

    @Override
    public NarrationPriority narrationPriority() {
        return current() instanceof NarratableEntry selectable ? selectable.narrationPriority() : NarrationPriority.NONE;
    }

    @Override
    public void tick() {
        if (current() instanceof TickingElement ticking) {
            ticking.tick();
        }
    }

    @Override
    public void stopDragging() {
        if (current() instanceof Draggable draggable) {
            draggable.stopDragging();
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/widget/WidgetEnvelope.kt
package juuxel.adorn.client.gui.widget

import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.Drawable
import net.minecraft.client.gui.Element
import net.minecraft.client.gui.Narratable
import net.minecraft.client.gui.Selectable
import net.minecraft.client.gui.screen.narration.NarrationMessageBuilder

/**
 * A wrapper for a widget (obtained by calling [WidgetEnvelope.current()][current]).
 *
 * Used for enhancing the widget with some functionality,
 * such as [pagination][FlipBook], [scissoring][ScissorEnvelope] or
 * [scrolling][ScrollEnvelope].
 */
abstract class WidgetEnvelope : Element, Drawable, Selectable, TickingElement, Draggable {
    protected abstract fun current(): Element

    override fun render(context: DrawContext, mouseX: Int, mouseY: Int, delta: Float) {
        val current = current()
        if (current is Drawable) {
            current.render(context, mouseX, mouseY, delta)
        }
    }

    override fun mouseMoved(mouseX: Double, mouseY: Double) =
        current().mouseMoved(mouseX, mouseY)

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean =
        current().mouseClicked(mouseX, mouseY, button)

    override fun mouseReleased(mouseX: Double, mouseY: Double, button: Int): Boolean =
        current().mouseReleased(mouseX, mouseY, button)

    override fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, deltaX: Double, deltaY: Double): Boolean =
        current().mouseDragged(mouseX, mouseY, button, deltaX, deltaY)

    override fun mouseScrolled(mouseX: Double, mouseY: Double, amount: Double): Boolean =
        current().mouseScrolled(mouseX, mouseY, amount)

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean =
        current().keyPressed(keyCode, scanCode, modifiers)

    override fun keyReleased(keyCode: Int, scanCode: Int, modifiers: Int): Boolean =
        current().keyReleased(keyCode, scanCode, modifiers)

    override fun charTyped(chr: Char, modifiers: Int): Boolean =
        current().charTyped(chr, modifiers)

    override fun isFocused(): Boolean =
        current().isFocused

    override fun setFocused(focused: Boolean) {
        current().isFocused = focused
    }

    override fun isMouseOver(mouseX: Double, mouseY: Double): Boolean =
        current().isMouseOver(mouseX, mouseY)

    override fun appendNarrations(builder: NarrationMessageBuilder) {
        val current = current()

        if (current is Narratable) {
            current.appendNarrations(builder)
        }
    }

    override fun getType(): Selectable.SelectionType {
        val current = current()

        return if (current is Selectable) {
            current.type
        } else {
            Selectable.SelectionType.NONE
        }
    }

    override fun tick() {
        val current = current()

        if (current is TickingElement) {
            current.tick()
        }
    }

    override fun stopDragging() {
        val current = current()

        if (current is Draggable) {
            current.stopDragging()
        }
    }
}
